internal/controller/buildpipeline/buildpipeline_adapter.go

```diff
@@ -111,6 +111,16 @@ func (a *Adapter) EnsureSnapshotExists() (result controller.OperationResult, err
 
 	expectedSnapshot, err := a.prepareSnapshotForPipelineRun(a.pipelineRun, a.component, a.application)
 	if err != nil {
```

Comment:
Thinking out loud, the `updatePipelineRunWithCustomizedError()` function seems to do much of the same things as the added lines.
To prevent code duplication, maybe something like this would work too here?
```
	if err != nil {
		result, err = a.updatePipelineRunWithCustomizedError(&canRemoveFinalizer, err, a.context, a.pipelineRun, a.client, a.logger)
		if h.IsInvalidImageDigestError(err) {
			return controller.StopProcessing()
		}
		return controller.RequeueWithError(err)
	}
```

Reply:
Read [updatePipelineRunWithCustomizedError()](https://github.com/konflux-ci/integration-service/blob/main/internal/controller/buildpipeline/buildpipeline_adapter.go#L357-L364) and I realized the current code has included `h.IsInvalidImageDigestError(cerr)` and return `controller.ContinueProcessing()` if `if h.IsMissingInfoInPipelineRunError(cerr) || h.IsInvalidImageDigestError(cerr) || h.IsMissingValidComponentError(cerr)` , so maybe we can just update this return in [controller.RequeueWithError(err)](https://github.com/konflux-ci/integration-service/blob/main/internal/controller/buildpipeline/buildpipeline_adapter.go#L115)? . Can you @jencull @dirgim help. to confirm it? Thanks!

Reply:
That's a good point, we currently ignore what's returned in [updatePipelineRunWithCustomizedError()](https://github.com/konflux-ci/integration-service/blob/main/internal/controller/buildpipeline/buildpipeline_adapter.go#L357-L364) and just check if an error was returned, which doesn't look right to me.

Since the function in question already returns the result that's expected by the controller, maybe simply doing this would work (untested, would need some testing to confirm):
```
return a.updatePipelineRunWithCustomizedError(&canRemoveFinalizer, err, a.context, a.pipelineRun, a.client, a.logger)
```

Reply:
@jencull are you fine with trying out the proposed change? Let me know if you need more info.

Reply:
One question: Does https://github.com/konflux-ci/integration-service/blob/main/internal/controller/buildpipeline/buildpipeline_adapter.go#L364 mean we will continue to call the other `Ensure**` functions in this controller thread? Because I notice https://github.com/konflux-ci/integration-service/blob/main/internal/controller/buildpipeline/buildpipeline_adapter.go#L348 will return  `controller.StopProcessing()` when meeting unrecoverable error `errors.IsForbidden(cerr)`